flatten_json.go

```diff
@@ -475,6 +496,60 @@ func (fj *flattenJSON) readLiteral(literal []byte) ([]byte, error) {
 	return literal, nil
 }
 
+func (fj *flattenJSON) skipBlock(openSymbol byte, closeSymbol byte) error {
```

Comment:
This func makes me nervous.  Normally when we're going through the buffer we use `fj.step()` and `fj.ch()`, is there a reason why this is different?

Also, nesting;   Suppose you have an array (arguments `[`, `]`) and one of the elements is an object, possibly deeply nested? Also, how about an object (`{`, `}`) and a member value is an array.

Reply:
> Also, nesting; Suppose you have an array (arguments [, ]) and one of the elements is an object, possibly deeply nested? Also, how about an object ({, }) and a member value is an array.

Nesting is covered here with the levels, you can see that we are skeeping most of `status.json` which contains nested objects with arrays.

I'll add specific tests for that, to validate skipping of objects & arrays.

> Normally when we're going through the buffer we use fj.step() and fj.ch(), is there a reason why this is different?

This is actually an interesting case! I wrote the code initially without the `step` and `ch` and when I used those methods the performance (time taken, not allocations) increased, talking about something like:
```
for {
  ch := fj.ch()

  // process

  if fj.step() != nil { return err }
}
```

So I tried to investigate this, it didn't make any sense, since this methods are inlined and cheap, so I did manual inlining and got no improvements.

Finally what I found that iterating over the `fj.event` with `fj.eventIndex` is slower than -
```
data = fj.event[fj.eventIndex:]
i = 0
// iterating from i until len(data)
```

I am not sure why iterating from the middle to from the start is slower, I was thinking about trying to check it with compiler explorer and see what's the difference .. I'll try that but I am not an assembly expert.

Reply:
Huh... agree, makes no sense to me. `data` is now a new `[]byte` against the same backing memory, maybe Go has some magic about indexing into `[]byte` when the index is small?  Anyhow, if this is repeatable, maybe we should have a fancier version of `step()` and `ch()` that rebases the slice and then all the parts of `flattenJSON` get the benefit?

Also, unless the improvement is dramatic I'm a little reluctant to have one piece of the big complicated function that arbitrarily has its own process for traversing the buffer.

Reply:
> Anyhow, if this is repeatable, maybe we should have a fancier version of step() and ch() that rebases the slice and then all the parts of flattenJSON get the benefit?

I was thinking about giving an option to have an iterator over the event that will hand you the current character and it might does what you are saying, but didn't want to introduce big change like that in this PR.

> Also, unless the improvement is dramatic I'm a little reluctant to have one piece of the big complicated function that arbitrarily has its own process for traversing the buffer.

The improvement was dramatic, I'll post benchmarks differences tomorrow and lets get a decision based on that.

Reply:
> The improvement was dramatic, I'll post benchmarks differences tomorrow and lets get a decision based on that.

Wow, given that ch() and step() are inlined, that is **really** weird (also they didn’t show up in profiling output).  Anyhow, there's a lesson hiding in here somewhere and once we figure it out we should give all of flattener the benefit. I agree, probably another PR for that.

Reply:
Update on benchmarks, I rewrote `skipStringValue` and `skipBlock` with `step` & `ch` -
```
benchmark                                                old ns/op     new ns/op     delta
Benchmark_JsonFlattener_ContextFields-10                 9062          12211         +34.75%
Benchmark_JsonFlattener_MiddleNestedField-10             10730         13619         +26.92%
Benchmark_JsonFlattener_LastField-10                     9570          12598         +31.64%
Benchmark_JsonFlattner_Evaluate_ContextFields-10         9487          12607         +32.89%
Benchmark_JsonFlattner_Evaluate_MiddleNestedField-10     11122         14021         +26.07%
Benchmark_JsonFlattner_Evaluate_LastField-10             9873          12893         +30.59%

benchmark                                                old allocs     new allocs     delta
Benchmark_JsonFlattener_ContextFields-10                 3              3              +0.00%
Benchmark_JsonFlattener_MiddleNestedField-10             4              4              +0.00%
Benchmark_JsonFlattener_LastField-10                     2              2              +0.00%
Benchmark_JsonFlattner_Evaluate_ContextFields-10         13             13             +0.00%
Benchmark_JsonFlattner_Evaluate_MiddleNestedField-10     13             13             +0.00%
Benchmark_JsonFlattner_Evaluate_LastField-10             9              9              +0.00%

benchmark                                                old bytes     new bytes     delta
Benchmark_JsonFlattener_ContextFields-10                 48            48            +0.00%
Benchmark_JsonFlattener_MiddleNestedField-10             64            64            +0.00%
Benchmark_JsonFlattener_LastField-10                     32            32            +0.00%
Benchmark_JsonFlattner_Evaluate_ContextFields-10         480           480           +0.00%
Benchmark_JsonFlattner_Evaluate_MiddleNestedField-10     488           488           +0.00%
Benchmark_JsonFlattner_Evaluate_LastField-10             312           312           +0.00%
```

I tried to write a representative benchmark - https://gist.github.com/yosiat/930426f4baea92b5f188203c6acb716b and it doesn't makes sense to me, I checked with https://godbolt.org/ and it looks that `step` instructions are the same as `for` code in it's relevant parts 🤔